### PR TITLE
Revisit the name mangling rule usage for the '$' character (method and filenames named differently)

### PR DESCRIPTION
**Pull request by pomadchin**
@jodersky also [mentioned](https://github.com/sbt/sbt-jni/issues/36#issuecomment-886038832) that it works [here](https://github.com/jodersky/mill-jnilib/blob/master/jnilib/src/jnilib/gjavah.scala) without extra rules for `$`. 

This PR is made to check what's going on with the header functions generation.

Closes https://github.com/sbt/sbt-jni/issues/36 (eventually)

### Context

JDK8 javah preserves a specific file naming when dealing with generating methods for Scala objects. It generates filenames for objects that contain `$` (all Scala objects)  as `package_ObjectName__.h` (mangles `$` into `__`). However, methods mangling is a bit different and for all methods it converts `$` into `_00024`.

Right now in the master branch, for the following `Adder.scala`:

```scala
package multiclasses

object Adder {

  @native def sum(term1: Int, term2: Int): Int

}
```

The following `multiclasses_Adder__.h` would be generated:

```c
/* DO NOT EDIT THIS FILE - it is machine generated */
#include <jni.h>
/* Header for class multiclasses_Adder__ */

#ifndef _Included_multiclasses_Adder__
#define _Included_multiclasses_Adder__
#ifdef __cplusplus
extern "C" {
#endif
/*
 * Class:      multiclasses_Adder__
 * Method:     sum
 * Signature:  (II)I
 */
 // An incorrect method name is generated!
JNIEXPORT jint JNICALL Java_multiclasses_Adder___sum
  (JNIEnv *, jobject, jint, jint);

#ifdef __cplusplus
}
#endif
#endif
```

With this PR the following `multiclasses_Adder__.h` would be generated:

```c
/* DO NOT EDIT THIS FILE - it is machine generated */
#include <jni.h>
/* Header for class multiclasses_Adder_00024 */

#ifndef _Included_multiclasses_Adder_00024
#define _Included_multiclasses_Adder_00024
#ifdef __cplusplus
extern "C" {
#endif
/*
 * Class:      multiclasses_Adder_00024
 * Method:     sum
 * Signature:  (II)I
 */
JNIEXPORT jint JNICALL Java_multiclasses_Adder_00024_sum
  (JNIEnv *, jobject, jint, jint);

#ifdef __cplusplus
}
#endif
#endif
```

cc @jodersky do we need to preserve the original jdk8 javah filename?
The alternative can be just to remove a special case around `$` and use the new namings generated by gjavah: `multiclasses_Adder_00024.h` which can be a bit more obvious (i.e. I would expect `multiclasses_Adder_00024` methods to be stored in a `multiclasses_Adder_00024.h` file).

TLDR; the whole difference in behavior is in this diff: https://github.com/pomadchin/sbt-jni/commit/ea4ecc86fea29c3b4b99bc4dfa5f25b42b6e78ac

### CODE REVIEW

**Comment by pomadchin** [comment]
Closed in favor of https://github.com/sbt/sbt-jni/pull/58 which I think makes more sense.
